js/packages/mobile-wallet-adapter-walletlib/android/src/main/java/com/solanamobile/mobilewalletadapterwalletlib/reactnative/SolanaMobileWalletAdapterWalletLibModule.kt

```diff
@@ -0,0 +1,386 @@
+
+        sealed class MobileWalletAdapterRemoteRequest(open val request: ScenarioRequest) : MobileWalletAdapterServiceRequest
+        data class AuthorizeDapp(override val request: AuthorizeRequest) : MobileWalletAdapterRemoteRequest(request)
+        data class ReauthorizeDapp(override val request: ReauthorizeRequest) : MobileWalletAdapterRemoteRequest(request)
+
+        sealed class SignPayloads(override val request: SignPayloadsRequest) : MobileWalletAdapterRemoteRequest(request)
+        data class SignTransactions(override val request: SignTransactionsRequest) : SignPayloads(request)
+        data class SignMessages(override val request: SignMessagesRequest) : SignPayloads(request)
+        data class SignAndSendTransactions(
+            override val request: SignAndSendTransactionsRequest,
+            val endpointUri: Uri,
+        ) : MobileWalletAdapterRemoteRequest(request)
+    }
+
+    private var request: MobileWalletAdapterServiceRequest? = null
+        set(value) {
+            field = value
+            value?.let { request -> sendWalletRequestToReact(request) }
+        }
+
+    private var scenario: Scenario? = null
+
+    private fun clusterToRpcUri(cluster: String?): Uri {
+        return when (cluster) {
+            ProtocolContract.CLUSTER_MAINNET_BETA ->
+                Uri.parse("https://api.mainnet-beta.solana.com")
+            ProtocolContract.CLUSTER_DEVNET ->
+                Uri.parse("https://api.devnet.solana.com")
+            else ->
+                Uri.parse("https://api.testnet.solana.com")
+        }
+    }
+
+    // Converts a react "ReadableArray" into a Kotlin ByteArray.
+    // Expects ReadableArray to be an Array of ints, where each int represents a byte.
+    private fun convertFromReactByteArray(reactByteArray: ReadableArray): ByteArray {
+        return ByteArray(reactByteArray.size()) { index ->
+            reactByteArray.getInt(index).toByte()
+        }
+    }
+
+    @ReactMethod
+    fun log(message: String) {
+        Log.d(TAG, "message from react: $message")
+    }
+
+    @ReactMethod
+    fun createScenario(
+        walletName: String, // our wallet's name (Backpack)
+        uriStr: String,
+        config: ReadableMap
+    ) {
+        val uri = Uri.parse(uriStr)
+
+        val associationUri = AssociationUri.parse(uri)
+        if (associationUri == null) {
+            Log.e(TAG, "Unsupported association URI: $uri")
+            return
+        } else if (associationUri !is LocalAssociationUri) {
+            Log.e(TAG, "Current implementation of fakewallet does not support remote clients")
+            return
+        }
+
+        // created a scenario, told it to start (kicks off some threads in the background)
+        // we've kept a reference to it in the global state of this module (scenario)
+        // this won't be garbage collected and will just run, sit & wait for an incoming connection
+        scenario = associationUri.createScenario(
```

Comment:
new scenario gets created each time. An edge case would be if react native caller tries to create a new scenario when one already exists. Should we return an error or do anything different in this case? Maybe we should rename the method to make it more clear that it will always overriding any previous scenario

Reply:
So, this method is never actually called by the JS dev right now (at least we don't explicitly expose it). JS side solely interacts with this under-the-hood when they call `useMobileWalletAdapterRequest`.

Should we allow the JS side to explicitly call this?